src/workerd/api/tail-worker-test.js

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2017-2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+import * as assert from 'node:assert';
+
+let resposeMap = new Map();
+
+export default {
+  // https://developers.cloudflare.com/workers/observability/logs/tail-workers/
+  tailStream(...args) {
+    // Onset event, must be singleton
+    resposeMap.set(args[0].traceId, JSON.stringify(args[0].event));
+    return (...args) => {
+      // TODO(streaming-tail-worker): Support several queued elements
+      let cons = resposeMap.get(args[0].traceId);
+      resposeMap.set(args[0].traceId, cons + JSON.stringify(args[0].event));
+    };
+  },
+};
+
+export const test = {
+  async test() {
+    // HACK: The prior tests terminates once the scheduled() invocation has returned a response, but
+    // propagating the outcome of the invocation may take longer. Wait briefly so this can go ahead.
+    await scheduler.wait(100);
+    // This test verifies that we're able to receive tail stream events for various handlers.
+
+    // Recorded streaming tail worker events, in insertion order.
+    let response = Array.from(resposeMap.values());
+
+    let expected = [
```

Comment:
This approach feels reasonably elegant – we test that onset/outcome events work for many different event types by reusing existing tests (much less overhead). By inserting into the Map based on the traceId, we keep all events sent to a given tail worker invocation together. There is some risk of race conditions affecting the order here, but since the individual workers are invoked by tests that are running sequentially this appears to work well in practice.

Note that this only checks the "event" portion of the TailEvent – the other components are less interesting/non-deterministic.

Reply:
Sounds reasonable to me